crates/sui-graphql-rpc/src/server/version.rs

```diff
@@ -0,0 +1,257 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use axum::{
+    http::{HeaderMap, HeaderValue, Request, StatusCode},
+    middleware::Next,
+    response::{IntoResponse, Response},
+};
+
+use crate::error::{code, graphql_error};
+
+const VERSION_HEADER: &str = "X-Sui-RPC-Version";
+
+const RPC_VERSION_FULL: &str = env!("CARGO_PKG_VERSION");
+const RPC_VERSION_YEAR: &str = env!("CARGO_PKG_VERSION_MAJOR");
+const RPC_VERSION_MONTH: &str = env!("CARGO_PKG_VERSION_MINOR");
+
+/// Middleware to check for the existence of a version constraint in the request header, and confirm
+/// that this instance of the RPC matches that version constraint.  Each RPC instance only supports
+/// one version of the RPC software, and it is the responsibility of the load balancer to make sure
+/// version constraints are met.
+pub(crate) async fn check_version_middleware<B>(
+    headers: HeaderMap,
```

Comment:
You should use a TypedHeader here instead in order to not require cloning all the unrelated headers

Reply:
Ah, I initially thought that `TypedHeader` only works with the header types built-in to `axum` but I see now that you can define custom ones, will do, thanks!